product_visible_discount_extension/models/sale.py

```diff
@@ -0,0 +1,206 @@
+                    record.temp_price = record.price_unit
+
+    @api.onchange('visible_discount', 'discount')
+    def _onchange_visible_discount(self):
+        """
+        Compute the price_unit based on the visible_discount value.
+
+        This method is triggered whenever the visible_discount is changed.
+        When the visible_discount is changed, the value of price_unit is
+        changed using this formula:
+
+            price_unit = (100 - visible_discount) / 100 * list_price
+        """
+        decimal_precision = self.env['decimal.precision']
+        precision = decimal_precision.precision_get('Account')
+
+        for record in self:
+            if record.product_id.id:
+                if record.product_id.list_price and record.temp_price is False:
+
+                    discount = 100 - record.visible_discount
+                    new_price = discount * record.list_price / 100
+                    record.price_unit = round(new_price, precision)
+                    record.temp_discount = record.visible_discount
+                    record.temp_price = record.price_unit
+
+                    if record.visible_discount < 0:
+                        record.visible_discount = 0
+                else:
+                    record.price_unit = record.temp_price
+                    record.temp_discount = record.visible_discount
+
+    def product_id_change(self, cr, uid, ids, pricelist, product, qty=0,
+                          uom=False, qty_uos=0, uos=False, name='',
+                          partner_id=False, lang=False, update_tax=True,
+                          date_order=False, packaging=False,
+                          fiscal_position=False, flag=False, context=None):
+        """
+        Setup the sale_order_line.
+
+        When the sale_order_line are modified by the module
+        product_visible_discount. It changes the amount for price_unit
+        and discount. We convert any change in price as a discount.
+
+        It convert discount, fixed price and mix of fixed price + discount
+        to a fixed discount amount.
+        """
+        res = super(SaleOrderLine, self).product_id_change(
+            cr, uid, ids, pricelist, product, qty, uom, qty_uos, uos, name,
+            partner_id, lang, update_tax, date_order, packaging=packaging,
+            fiscal_position=fiscal_position, flag=flag, context=context)
+
+        decimal_precision = self.pool.get('decimal.precision')
+        precision = decimal_precision.precision_get(cr, uid, 'Account')
+
+        if 'price_unit' in res['value']:
+            new_price = None
+
+            if res['value']['discount'] == 0:
+                # If the pricelist has a fixed price it doesn't set
+                # a discount
+                product_obj = self.pool.get('product.product')
+                product = product_obj.browse(cr, uid, product, context)
+                price_list = product.list_price
```

Comment:
@llacroix Are you considering the pricelist in the SO or just the product's base price?